### DISCUSSION TITLE
Missions

### DISCUSSION THREAD

**Discussion by alexanderson1993** [Thorium-Sim/thorium-nova#47] [Ideas]
This discussion should mostly center around purely automated missions. Further discussion about the role of Flight Directors is being held here: https://github.com/Thorium-Sim/thorium-nova/discussions/46

Let's start from first principles. Thorium Nova's approach to missions is ultimately what sets it apart from other bridge simulators and justify its existence. Story and narrative should be the focus here, although action/arcade elements certainly should be possible. People should choose to play and replay Thorium Nova both because they enjoy the experience of using the controls _and especially_ enjoy the story.

There's no reason why someone can't just hop into a spaceship and travel around the universe, but without there being anything to do, it could get very dull. On the other hand, just being given assorted tasks with no substance or meaning makes playing the game a drag.

Here are two videos from Extra Credits which illustrate this nicely:

https://www.youtube.com/watch?v=otAkP5VjIv8
https://www.youtube.com/watch?v=ur6GQp5mCYs

One convenience that we might have: The lore could be written such that the player is part of some organization, military or otherwise, that assigns them their missions, so the crew doesn't have to wander around space aimlessly trying to find things to do. (Although, it should be possible to assign side-quests as the crew interacts with the universe.)

So, missions serve a few purposes:

- Provide context and background to the crew about the universe.
- Provide the crew with some overarching objective - something obvious to focus their efforts on - along with a place to start to accomplish that objective.
- Make it easy for the crew to know what to do next.
- Give the crew a reason to explore the universe, go places, do things, talk with people, etc.
- Provide a vehicle for telling a a story.

I went through all of the missions that I could remember from the Utah space centers, and  they primarily deal with the crew doing some variation on these actions:

- Travel to location
- Speak with person, negotiate, eavesdrop
- Get an item/cargo
- Give an item/cargo
- Investigate
- Find something or someone
- Escape/Hide
- Fight/Destroy
- Escort a ship/fleet

Going through the timelines for some Thorium Classic missions, these are the actions which are fired to set up the different parts of the mission:

- Change what's on the viewscreen (likely not relevant to Thorium Nova, since the viewscreen should be fully automated)
- Change the listed mission objectives
- Send messages to the bridge from someone on the ship
- Send a long range message to the crew from another ship
- Create sensor contacts and targeting contacts
- Send special sensors data
- Start a short-range communication signal.
- Damage or repair a ship system
- Play special sound effects
- Change the lighting to a specific setting

It seems most of the actual mission progression and storytelling is done by the Flight Director in Thorium Classic, not the controls.

It should be possible for the mission system to know how to assign objectives like these to the crew, and how to know when the crew has accomplished these objectives. I imagine a lot of that will fall upon the mission writer, but the system should make it as easy as possible.

### Linear or Branching

Many of the missions I'm familiar with take a very linear approach to storytelling, since that makes the flight directing experience more predictable. Without a Flight Director, though, branching storylines are very possible. That tells me that we need the ability to to branching storylines. This also lends itself well to "Choose your own Adventure" style missions, where different choices affect the overall outcome of the mission, which increases replayability.

Even still, we need an easy way for mission writers and Flight Directors to visualize the progression of the mission, branching or otherwise. That's something I have yet to figure out.

### Campaigns

One thing that would be awesome to build in from the get-go is campaigns of missions, or missions that run from one into the next. This could be done as part of a single, long running flight (with pauses between sessions). This implies the ability for flights have multiple simultaneous missions (although it's most likely that one mission's end will trigger the other mission to start).

Or each mission could be a new flight, but the crew can save and load the ship they flew on to provide some continuity between flights. This requires a feature for saving the player ship's state to be used elsewhere, and implies RPG style features, like buffs, upgrades, and leveling up.

Then again, it might also be desirable to use a different ship with different capabilities for each mission. Should there be some way to block a crew from playing a mission later in the campaign if they haven't completed the previous missions?
 
Regardless, I envision campaigns working like Gloomhaven. You can see in this diagram how each of the scenarios leads into the others, with some dependencies on others, some being blocked if you choose to do things a certain way, and lots of choice as to what your next objective is.
<img width="1002" alt="Scenario Flowchart" src="https://user-images.githubusercontent.com/6558157/137498653-9b32903f-0fdc-4b16-9d28-a989e2526ffc.png">

### How Missions Should Work

Honestly, I don't know. There are so many ways we could model this. To summarize the goals:

- Clear objectives for the crew
- Some way to tell if the objective has been completed
- Completing tasks or triggering event listeners causes the mission to progress
- Easy to understand tools for creating and composing missions
- Campaign missions
- Branching storylines

Thorium Classic used timelines to create a purely linear approach to missions. It's possible we could take advantage of this, where missions are comprised of multiple timelines that activate throughout the mission. Creating missions would then turn into creating these isolated chunks of timeline that are connected through the mission.

Then again, a more objective-based writing approach might make more sense - When the crew is given an objective, the system also initializes all of the things needed to accomplish that objective, like creating entities to destroy, assigning conversation trees to NPCs, placing cargo where it can be found, and giving behaviors to fleets of ships.

Something tells me neither of these approaches are quite right, and as we talk through this and build it, a better solution will emerge.

**Comment by Xansta** [comment]
If there's a way to answer this question incrementally, that would be good. I don't know that we'll have all the answers up front. 

I also think that we might be able to use mini-missions to create tutorials ala Empty Epsilon. For each card, there can be a mini-mission that serves as the tutorial.

Allow me to demonstrate my ignorance of Thorium cards with an example. 
**Card:** Open/Close the exterior cargo bay door. 
**Mini-mission:** Load Fuel Cells Into the Cargo Bay. 
_Prompt:_ text message/bubble appears with text, "The Argyle Consortium will be delivering fuel cells. They need to be loaded into the cargo bay. When the fuel cells arrive, the Argyle Consortium Delivery bot will request access to the cargo bay. You need to open the door at that time." There is a close or accept button in the message/bubble/dialog box.
_Timer_ for x time during which opening the cargo bay door responds with something like, "doors open" and/or "you were not given the authority to open the doors yet"
_Timer expires_
_Prompt:_ text message/bubble: "[Argyle Consortium Delivery Bot] Request access to cargo bay"
_Wait for client to open the cargo bay door_
_Indicate cargo delivered via text message, graphic, sound (any of the above)_
_Prompt:_ "Now that the cargo has been delivered, please close the cargo bay door so that we can repressurize the cargo bay"
_Once the door is closed, indicate mission success._

Creating mini-missions like this exercises the card and shows how it's used. If there are several cards combined (like the exterior cargo bay door, the interior cargo bay door and the environmental controls of the cargo bay), then a mini-mission might be a bit more complicated with more failure points (forgetting to depressurize the bay before the cargo is delivered, opening both the interior and the exterior doors at the same time, etc.)

I know that does not answer the overarching question about how missions should work, but if we start small, we can see how different aspects interact, build up a library of missions and get a better idea of how an overall strategy could/should be implemented.

**Reply by alexanderson1993** [reply]
I actually drafted a different version of this discussion which had four different levels of "mission", where your mini-mission fits in as one of the levels.

- Tasks - individual actions that an individual crew member can do that can be tracked by the system. Opening the cargo bay door and closing the cargo bay door are both separate tasks.
- Objectives - Groups of coordinated tasks that one or many crew members do together. I consider your mini-mission to fit under this category.
- Timelines - A series of tasks and objectives that feed from one to another. "Load the fuel cells into the cargo bay" leads to "Bring the fuel cells to Arcturus II" leads to "Unload the fuel cells". All three of those together create a small timeline.
- Missions - A briefing to give context along with a number of timelines in series and parallel - branching and merging as the mission progresses.

I didn't publish that draft because I felt I didn't have a good example of how it all plays together. Your mini-mission helped me articulate it.

Also, it would be kind of novel to execute training by dynamically starting up mini-missions/Objectives specific for each card and station.

**Reply by astrolamb-gaming** [reply]
I think your organization of the four levels of a mission is an excellent one. It's likely that we could set up mission definitions using these as building blocks.

**Comment by alexanderson1993** [comment]
One thing that is _crucial_ which I didn't emphasize enough is that mission writers could be anyone. Obviously, contributors to this project will help create the built-in missions for Thorium Nova. But anyone should be able to do it as well. Flight Directors, ambitious crew members, families and friends wanting to create their own personalized stories, or professionals creating missions for their con setup. The assumption is that anyone should be able to create a mission with no in-person training or pre-experience - with a combination of intuitive controls and good documentation.

**Reply by astrolamb-gaming** [reply]
Regarding setting up a mission editor for new people, I have a bit of a vision for that. 

We would have a node-based visual "scripting" system with several levels, essentially one for each of the [levels you've described](https://github.com/Thorium-Sim/thorium-nova/discussions/47#discussioncomment-1484276). 

For the mission level, there would be several template timelines for mission writers to use, with the ability to customize, set up custom templates, or wholly new templates.

This motif would exist for each of the levels: timeline level would have objectives, objective level would have tasks. I think it'd be helpful for timelines to be able to be used within other timelines though.

For the rest of this post I will use the terms level and node, where if level is a timeline, then node would be an objective.

You could build a level that has several nodes in series or parallel depending on the situation. The user should also be able to define trigger and listener nodes for events, which could be located within any of the levels. The events would allow effects to happen on things outside of the immediate level, for example, blocking a node path in any given level.

We'd also have timer and delay nodes, which would allow specific timing.

**Comment by astrolamb-gaming** [comment]
Couple thoughts based on our meeting on Oct 19th:
Mid-mission, the graph should show where the the player(s) are in each timeline, probably with some sort of colored outline for each node in a graph - possibly green for in-progress, red for failed, blue for succeeded, and grey for not started.

We discussed the [flume system](https://flume.dev) during the meeting, which from a user interface standpoint is much aligned with what I envision for much of the flight director interface. What I'm not totally certain about is how we will take these nodes and utlize them as a timeline. Flume can be used as a programming language, and there will be times where flight directors will use it as such (particular in initialization stages and other trigger-based events), but we have to remember that it will be used for tracking a series of steps as well, and I'm not sure how that will work exactly. (@alexanderson1993 if you have a vision for implementing that, please enlighten me!)

**Reply by alexanderson1993** [reply]
I thought about this a lot last night, and I think I came up with something. 

The mission graph ultimately needs to do two things:
- Check to make sure certain actions have been performed.
- Trigger changes in the simulation.

I think it _still_ makes sense to use the mission graph as a giant function. We just run the function on a loop. Every 1 second or so, it executes the mission graph, checks to see if actions have been performed and triggers the changes if necessary. To avoid executing steps multiple times, and to make future execution faster, it marks nodes as "completed".  Completed nodes memoize their return value (if any) and don't execute again.

Flume is awesome, but it doesn't cover all of the features we want. For example, you can't encapsulate several nodes together into a single node, and some features I want around drag-selection and copy/pasting don't exist. Those features will be necessary to make it easier for mission writers to refactor the structure of their missions. We'll likely want to fork Flume to make the modifications that we want.

There's also part of me that wonders if we can lean into the "Task", "Objective", "Timeline", "Mission" abstraction to provide a clearer user experience for the Flight Director in the middle of the flight than having them look right at the mission graph. Perhaps, by organizing missions into timelines, the Flight Director sees the active timelines (the ones that have started that are not yet completed) and can expand them to see the current objectives, which they can expand to see the current tasks.

There also need to be some specialized nodes which can be used throughout to provide data and context. These include
- The variable import/export nodes
- Nodes that can find specific ECS entities. For example, a "is ship" filter node, a "faction" filter node, and a "within radius" filter node could be used together to grab a rebel ship within 200,000km of the planet Alpha Centauri A. 
- Nodes that trigger inputs (server mutations) or that directly manipulate entities.
- Some nodes should be able to access the list of all of the inputs that happened since the last time the mission graph was executed, to know if a certain event has happened.

As far as actually executing on this feature, I think it won't happen for a little while. We need to build out some of the crew controls first, so there's actually something for missions to be built around. Then we can start small, just by implementing the task system (which allows the system to register tasks that crew members need to do and automatically mark them as complete when they are done). I think tasks could be implemented with a node graph as well, which should be fun. 

Then, we can start adding complexity by adding in the objectives, timelines, and then finally full missions. This is a huge endeavor, but I think it's possible, and what's more, I think it can be a great mission writing experience for beginners.

**Reply by alexanderson1993** [reply]
In case anyone was interested, this is the raw JSON output for a simple Flume node graph. 

```json
{
  "mh60qMEhcB": {
    "x": 190,
    "y": -150,
    "type": "homepage",
    "width": 170,
    "connections": {
      "inputs": {
        "copyrightYear": [
          {
            "nodeId": "TXPXrEdMOG",
            "portName": "number"
          }
        ],
        "title": [
          {
            "nodeId": "gd2sqwkdi4",
            "portName": "string"
          }
        ]
      },
      "outputs": {}
    },
    "inputData": {
      "title": {
        "string": ""
      },
      "description": {
        "string": ""
      },
      "showSignup": {
        "boolean": true
      },
      "copyrightYear": {
        "number": 0
      }
    },
    "root": true,
    "id": "mh60qMEhcB"
  },
  "TXPXrEdMOG": {
    "id": "TXPXrEdMOG",
    "x": -176,
    "y": -111.5,
    "type": "number",
    "width": 160,
    "connections": {
      "inputs": {},
      "outputs": {
        "number": [
          {
            "nodeId": "mh60qMEhcB",
            "portName": "copyrightYear"
          }
        ]
      }
    },
    "inputData": {
      "number": {
        "number": 2013
      }
    }
  },
  "gd2sqwkdi4": {
    "id": "gd2sqwkdi4",
    "x": -181.74999237060524,
    "y": -293.4999847412106,
    "type": "string",
    "width": 200,
    "connections": {
      "inputs": {},
      "outputs": {
        "string": [
          {
            "nodeId": "mh60qMEhcB",
            "portName": "title"
          }
        ]
      }
    },
    "inputData": {
      "string": {
        "string": "Hi there!"
      }
    }
  }
}
```

**Reply by alexanderson1993** [reply]
I'll be completely honest, I'm not totally sold on the idea of a node graph, as cool and powerful as it might be. It might be overkill, overly complicated, and over engineered. I wonder if there are alternative ways to represent this logic (aside from code) that might be _easier_ to configure and easier to represent to a flight director in the middle of a mission.

Another thing to consider, especially in light of having missions that are only written with code: We should think about whether representing the entire timeline to the Flight Director in the middle of a flight should be out of scope. It adds a lot of complexity and challenges, and might not be necessary. For example, if we make it so the mission can provide prompts to the Flight Director at the appropriate times, and the Flight Director can also see what the crew's current objectives are, that might be enough.

I'm just thinking of how challenging it would be to display the overall mission to the Flight Director if it were written in code. It's not impossible, but it's really hard to think through. I think, instead of exposing the Flight Director to the underlying data structures, we apply abstractions on top (like I described above) which only show the Flight Director what the absolutely need to know to do their job.

This again goes back to "What is the role of the flight director?" (https://github.com/Thorium-Sim/thorium-nova/discussions/46) and how involved in the inner workings of the system they should be.

**Comment by alexanderson1993** [comment]
Okay, okay okay. New proposal. We still use a graph editor, but its role is much smaller. Instead of building out entire graphs for the whole mission, we only build graphs for smaller pieces, that are then inserted into timelines. Timelines will be much more similar to the Thorium Classic timelines, where there are steps that are progressed through.

Missions are still composed of timelines, both parallel and serial. Timelines are composed of discrete timeline steps which are stepped through as the mission progresses. 

Timeline steps are comprised of 
- Inputs (mutations to the simulation, like sending a long range message to the crew or spawning a ship),
- Tasks (automatically verifiable actions assigned to individual crew members), 
- Objectives (Essentially event listeners that are trying to see if the simulation is in some state, like whether certain cargo has been moved to a certain ship or whether a specific ship has been destroyed)

Tasks and objectives can be assigned actions which are performed once the task or objective is complete, like automatically advancing to the next step in the timeline, starting a new timeline, setting a variable, or triggering more inputs. By default, timeline steps will automatically advance once all of the tasks and objectives in the timeline have been completed, but of course any objective and task can directly advance to the next timeline step. Delay nodes can be added anywhere in the process, and can be set to wait a certain time after all tasks and objectives in the timeline step have been completed before automatically advancing. 

(An important distinction: Naming things is hard. The term "Objective" here shouldn't be thought of as "Mission Objective". For example, if the crew's overarching mission objective is "Deliver cargo from Septimus II to Rigel 4", there would likely be several timeline steps, each with their own objective. The first might be "Course is set to Septimus II and engines are activated", which would advance to the second timeline step where the objective is "Player's ship enters the Septimus II system" - neither complete the overarching mission objective - they just serve as milestones to advance the timeline)

Where does the graph editor come in? Objectives and tasks are still created using the graph editor. Each objective and task is a small graph to determine whether the task or objective has been completed. Maybe we keep the four states (inactive, in-progress, success, failure) and make it so post-completion actions can be different based on whether the objective was a success or failure.

The nice thing about this is that we can easily show the Flight Director what's going on in the simulation. They can see what timelines are active, step through the timeline steps, see the current Objectives
(Also, as I wrote the above, I thought to myself "Wow, objectives and tasks are basically the same thing... can we simplify this at all?" Maybe we should just combine those two concepts together, or separate the concept of "Tasks" from mission writing altogether and only worry about whether the simulation is in a certain state or certain inputs have occurred.)

It should be obvious that this is a complicated feature and will go through many iterations, both in the design and discussion phase and in the development phase. So, bear with me if my backpedaling on the mission graph concept is frustrating to you.

Let me know if you like it, hate it, or think if it should be any different.

**Reply by astrolamb-gaming** [reply]
I'll start with Objectives vs Tasks, since that's a much smaller thing.... Since objectives are merely a set of tasks, it's not a stretch to think of them as timelines in their own right, especially as we intend to support timelines within timelines, potentially ad infinitum. So, sure, we can toss that level. This would also alleviate confusion between how we've been using it and mission objectives. If we do keep it, we could rename it to "milestone" instead.

Also real quick: might be best to have a fifith "skipped" state for timelines, for when a flight director and/or the crew decide to skip a particular timeline.

As for timelines as a whole:
While we were reviewing the old timeline system in Thorium Classic, I actually really liked the simplicity of the timeline view. with a scrollable list of events that should happen in order. While there is a need for changes to allow asynchronous timelines, I might argue that Classic's representation of a timeline is more user-friendly than the graph system. As a somewhat relevant side-note, I think that the vertical representation of timelines is more intuitive than the more horizontal view of the node graphs. There's a reason that folder directories are top-down rather than left-right. Incidentally, a [folder directory structure](https://cdn.dribbble.com/users/1916867/screenshots/7718583/media/b93175d2639599efc6b6a640718f4927.png) might be a good way to handle visualizing nested timelines.

That said, the ability to move the stuff _in_ a timeline is important, as are asynchronous (parallel) sub-timelines. There are two possibilities here: either timelines by default perform their steps in order, or they can perform them in any order. This could be merely a flag as part of each timeline (probably the simplest and best way to deal with this), but with some fanagling, I could see asynchronous timelines being an emergent property of nested timelines.

Really the only potential difficulty I see in a vertical representation of timelines is their inputs and outputs. The horizontal, node-based system is pretty nice for allowing an easy representation of more dynamic events for which events in one can influence another without lots more defined actions or a flight director. 

As an example, say there is a series of negotiations, where depending on result of the negotiations, the crew is asked to transport x amount of cargo to a location. x in this case might be defined in one timeline and needed in another. Such interactions between timelines would be easily followed if a node-based visual system were used. 

So, all that said, I think that both systems can be used. The list/folder structure variant would be great for overviews - a "here's what's planned to happen relative to each other" view, while the graph/node structure is better for modifying things.

**Reply by alexanderson1993** [reply]
> milestone

So let it be written: `milestone` -> `timeline step` -> `timeline` -> `mission`

> might be best to have a fifith "skipped" state for timelines, for when a flight director and/or the crew decide to skip a particular timeline.

I actually think there only need to be _three_ states: inactive, in progress, and completed. 

What is the purpose of "success" and "failed" states? My thought was they provided two separate paths for starting new timelines, but having just two possible outcomes of a timeline seems limited. Instead, there are three ways a timeline could start a new timeline:
- It's set up to automatically advance when all of the milestones are completed
- One of the milestones in the timeline triggers a new timeline to start when the milestone completes
- The Flight Director chooses to skip a timeline altogether. All of the milestones are marked inactive, none of the timeline steps execute, but the next timeline in the series activates

The "success" and "failed" states are only helpful for timeline creation on auto advance, but how does it know if the timeline was a success or failure in the first place? In which case, why don't you just have the milestone trigger the new timeline?

> x in this case might be defined in one timeline and needed in another.

I think this could be handled by setting and accessing variables across the mission.

**Reply by alexanderson1993** [reply]
Here's an example UI from Dream Flight Adventures that shows how I envision the Mission and Timeline portions might work.

https://user-images.githubusercontent.com/6558157/138187456-6c3c82d7-ab74-4f09-a621-0e230153511f.mp4

Notice that its very Flight Director guided - no automation at all aside from automatically progressing to the next timeline step if the steps are linked together.

Also notice that there's no possibility for doing parallel timelines. It's just a single, linear timeline. but the Flight Director does have the ability to modify the timeline order, execute timeline steps in any order, link and unlink timeline steps, etc. mid flight. Double clicking on the timeline steps opens up a window showing the actions which happen when the step is activated. For our purposes, this would instead show Inputs and Milestones.

<img width="455" alt="Screen Shot 2021-10-20 at 7 54 49 PM" src="https://user-images.githubusercontent.com/6558157/138188088-5b52d298-7613-4dcc-8e16-4d283136a93d.png">

Color coding the timeline steps would be an added bonus.

I think building on a UI like this would be awesome, with just a few modifications to allow for branching timelines and automation.

**Reply by astrolamb-gaming** [reply]
Forgot I was going to comment on this. I do really like how clean this UI seems. I'm not sure how branching timelines would look with this sort of setup. Perhaps if this was the timeline view, you could open up sub-timelines, which would show the same UI but for the sub-timeline instead? Then a back button somewhere to go back to the parent timeline?

**Comment by Xansta** [comment]
When I started reading, the first thing I thought was that "objective" seems to be independent of the abstraction being described. Then I saw the consolidation of task & objective along with the discussion of the incorporation of the term "milestone" (something I was going to suggest).

Right now, I don't think there's a good way to visually represent simultaneous overlapping objectives or timelines. As discussed, there may be an overarching objective consisting of several smaller objectives. These may be timelines of varying length, with or without interdependencies.

One of the things that I use quite a bit in scenarios for Empty Epsilon is having the story elements influence the way friendly, neutral and enemy ships behave. Some of these are built in to EE, others I have to write in the applicable conditions. Examples: 

- Built in AI which can be selected for an NPC ship via script and/or in the GM screen
  - Defend a target or location
    - Travel to, then circle the target (a station, another ship, etc.) or location defensively until an enemy approaches within a certain distance, then go out to intercept them.
  - Dock 
    - Travel to the location and dock unless an enemy approaches, then if you're armed, attack, if not, evade.
  - Roam
    - Look around for enemy targets to attack. Travel in a random direction if nothing is within sensor range.
  - Stand Ground
    - Do not move. Shoot any enemy that approaches
  - Fly To
    - Travel to designated location, attack enemies in range along the way
  - Fly Blind To
    - Travel to designated location, ignoring any enemies along the way
  - Fly in Formation (scripting only, not available in GM screen)
    - Ship is given a distance and coordinates to use to stay positioned relative to the leader's position
- Simple behaviors I've written in
  - A station spawns enemy ships periodically until that station is destroyed. Different orders for the ships may be: 
    - Roaming
    - Attack player ship
    - Attack designated base
    - Fly to designated location
  - A ship or a group of ships flies a patrol pattern consisting of multiple points
    - This can cover more area than the built in defend target or location AI
    - Points can vary as the scenario progresses
  - Send harassing messages to player ship while chasing them
    - Each set of messages is triggered by a particular circumstance
    - The individual message is selected at random from a list of messages
  - Spawn enemy ships based on the relative strength of the player ship(s) present
    - Each enemy ship type is given a numeric weight
    - Each player ship is given a numeric weight
    - The script determines the strength of the enemy fleet using difficulty setting and ratio of enemy to player

I provide these as examples of things I'd like to see available to the mission script writer either directly or through flexible use of the tools we provide. From these examples, you can see that there is a strong emphasis on ship to ship interaction, mostly combat in Empty Epsilon. I also wanted to demonstrate that some of these behaviors don't necessarily fit into a linear story telling format. Ships on patrol will simply be on patrol. If they are enemies and the players approach, they'll attack. The story is somewhat implied: the patrolling fleets believe they have something to protect and are thus doing their protective duty. 

The mission story could directly reference this behavior with something like, "Be wary of the Greton area. Their ships will attack if you get too close," or "You must get through the Greton area to reach Frilton 6. Avoid or engage as the circumstance warrants," or "We are sending five fighters to accompany you through the Greton area. Command them as necessary. Engage or avoid as you see fit. Remember your primary mission is to rendezvous with Excalibur at Frilton 6."

I don't necessarily see these examples as strictly a linear story. The ship behaviors make up the background of the universe, at least initially. In a fully automated mission scenario, the ships may change their behavior based on circumstances such as, after a certain amount of time, all enemy ships move to attack the players or their primary base, or once the players approach a location or station, the enemy ships flee, or attack or self-destruct. If a GM or flight director were in control, the GM could direct these ships based on text or verbal interaction between the players and the NPC captains or admirals (yes, I know this discussion is designated primarily for automated missions rather than GM directed missions).

In summary, what I'm suggesting is that some story elements may be provided via AI associated to NPC ships. Also, I provided the Empty Epsilon AI and GM tool examples as things that I think are valuable and that we should consider emulating. I'm not saying we have to, just making the suggestion. Finally, AI based elements may not have a good visual representation in a GUI based mission writing tool, nor may overarching or semi-independent story elements that may be more associated with the environment rather than directly to the story.

**Reply by alexanderson1993** [reply]
It think everything you mention should be possible using tools that we create. A separate discussion should be started about ship AI and all the different behaviors we want them to be able to exhibit. I've got a list somewhere that I've made which has a whole bunch that we can try to implement.

I also think that there might be some ceding of control that has to take place compared to Empty Epsilon. For example, EE operates on a 2D plane with unrealistic dimensions. The goal for Thorium Nova is to operate in a 3D world with realistic distances and sizes. As an example, I imagine using a visible map GUI to input the points (in kilometer units) that a fleet of ships patrol along would be a little easier than manually entering those points.

As an aside on that point, it's entirely possible that doing realistic distances and sizes is a terrible idea, both for action gameplay and for storytelling. Like, having ships patrol even around a planet, let alone an entire solar system, is just unfeasible.

In other words, I think everything you suggest should be possible, but I also think the style of mission writing will be different between EE and Thorium Nova.

**Comment by Xansta** [comment]
Indeed, one of the guiding principles of EE (particularly regarding its simulation model) is fun over realism. And, yes, EE is very 2D in its play model. 3D comes in the visuals only. I'm not necessarily advocating a change in philosophy away from realism, just noting and agreeing with the differences.

Specifying a destination in the GM screen for EE is done with a click. Of course, the GM screen is 2D, so that makes it simple. It would be harder to get a 3D GUI model working to accept coordinates when your typical input devices (mouse, trackpad, touchscreen, keyboard nub) are all 2D in nature. You could do it in stages: specify the plane angle (typically the ecliptic where all the orbiting bodies are), specify how far off the plane (above/below), then click. 

I think we're all mostly Kahn-2D in our perceptual assumptions.

And yes, I agree, the mission writing style will be different. I'm just lobbying for my favorite aspects from EE early.

**Reply by alexanderson1993** [reply]
> fun over realism

Agreed agreed. If it wasn't clear before, I'm more than happy to abandon this notion of realistic sizes and distances if it's no fun with them. I think it's worth a shot though.

> 3D GUI model

Thorium Nova Prime solved this by having the FD screen be a top-down view, much like EE, but you can choose the y-position of the ships when you spawn them and give orders. A few links from the blog:

https://thoriumsim.com/blog/update-16-map-updates
https://thoriumsim.com/blog/the-flight-director-star-map

> I'm just lobbying for my favorite aspects from EE early.

Keep it up! Your perspective is really valuable.

**Reply by astrolamb-gaming** [reply]
Regarding the map: I really like how the galaxy in Stellaris is rotatable and zoomable. Ditto for Kerbal Space Program. I think that sort of functionality would be huge from a gameplay perspective, since no other bridge sim that I'm aware of has that. 
As for placing/moving objects in 3D, KSP allows you to use widgets for translation and rotation just like in the unity or unreal editors, or any 3d modelling software. That functionality for a flight director would be great.

**Reply by astrolamb-gaming** [reply]
Regarding realistic distances, etc: I vote we keep them. However, if it takes a few minutes to travel from one place to another, I might suggest some sort of "skip warp/jump" feature (this is a great aspect to Pulsar: Lost Colony; it gives the crew time to step away for a few minutes if they want, and when they're ready they can skip warp and go right back into the next part of the story).

**Reply by alexanderson1993** [reply]
> rotatable and zoomable star map

I've done it before, we can do it again. Everything you suggest is possible, although I think having a 'fixed top-down' view option for the crew would be helpful, just in case they get lost.

**Comment by Xansta** [comment]
This may warrant its own topic, but moving the ship from place to place depends on immediate circumstances. Specifically, navigating between two star systems or even two planets takes a different skill set than maneuvering the ship during combat to bring weapons to bear or avoid inbound missiles. The first is more strategic, the second more tactical. I think that Artemis and Empty Epsilon lump both of those skills into the helm station. 

What's supposed to happen is that when going to some point beyond the short range sensors, Helm awaits instructions from the Captain or from Science or Relay regarding the destination, then some kind of direction is given like, "set course 315." Sometimes the Captain will say something like, "take us to Deep Space 17" and either Helm, who has a second screen up showing the destination, will set that course or Science or Relay will provide direction. In Artemis, each player can switch over to the Long Range Sensor (LRS) view to get the bigger picture, which looks roughly the same for Science, namely a 5x5 grid. Empty Epsilon pushes the crew to work together a bit harder because Science can only see to the edge of their sensors (typically 30 units), while Relay can identify friendly bases and ships through their long range telemetry. Relay can set a waypoint which shows up on Helm and Science screens and say something like, "Deep Space 17 is at waypoint 3". If Science can see it, they can click on it and say, "Deep Space 17 is at bearing 242." In either case the dynamic of the crew working together towards a goal is emphasized.

In a combat situation, helm cares more about positioning the ship for best weapons application or the best evasion of enemy fire than setting up for long range travel. This is where 3D vs 2D can become interesting/challenging. Artemis has some 3D in the way it handles combat, but to the Helm, the view shows a top down view of the ship with lines showing where the beam weapons hit (arc and range). The 3D aspect for helm comes when they choose to go up or down on the plane. This usually gives the weapons officer a chance to hit the other side of the ship if desired in manual targeting mode. EE has nothing like this for tactical combat - it's all 2D all the time. As mentioned previously, 3D is in the visuals for EE, but not in the game play model.

If we want something that the other bridge simulations don't have (except maybe SNIS), I suggest that beam weapon placement on the ship and subsequent firing incorporate the 3rd dimension. Thus, instead of an arc of coverage, we might have a cone of coverage. Helm might be constrained to roll the ship to bring additional beams to bear on a target rather than just rotate the ship. To help visualize, think of the quad cannons on the Millenium Falcon: there was one on the top and one on the bottom.

With that in mind, I also suggest that there be turrets for beam weapons rather than just plain arc coverage. EE has something like this today, but the turrets rotate automatically to cover the target on which the weapons officer has locked on. I suggest that beam weapons have the chance for target lock to fail requiring manual aiming of the beams and that the position of the ship in three dimensions along with the position of the turret determines the effectiveness of the beam weapons.

If I were a Klingon, Kralien or Kraylor approaching a ship in the Artemis or EE universe, I would approach from above or below the ship since the beams on those ships only seem to point along the plane that they are flying. 

To summarize, the skills needed to get the ship from point to point over long distances differ from the skills it takes to maneuver the ship in a combat situation. From what little I've read of the previous Thorium incarnations, tactical combat decisions were minimal or non-existent. We may want to consider devising two kinds of piloting roles.

One other related item: launching fighters or shuttles. They may or may not be used in combat. They will certainly be more maneuverable than the mother ship. Ok, enough mynock chasing for now.

**Reply by alexanderson1993** [reply]
Yeah, this should totally be its own topic. That said, I like where you're going with this. I absolutely agree that Nova could do better than its predecessor at tactical maneuvers.

**Comment by BrackenFunk** [comment]
I feel like missions are as linear as they are due to time frame. That's always been the case. Stories are written to fill "x" amount of time, and so therefore it's more linear to ensure we get to the resolution.

As Thorium Nova is going to be open to a variety of hosts and forms of operations, I understand the need for the open-ended exploration "quest"-style of operations. But there does need to be the ability to set "rules" and reign in how open-world/open-ended the scenarios are, right?

What I guess I'm trying to ask/say is that there needs to be a way to craft rules that limit the power of the crew for us commercial centers. We may be able to offer the open ended missions (and save them in some way) as well - but we also have normal operations that will require us to be able to limit where and how a crew can go. Does that make sense?

**Reply by alexanderson1993** [reply]
> Stories are written to fill "x" amount of time, 

Coming from the perspective of D&D and other tabletop RPGs, it's possible that some stories take much longer or shorter than the story writer intended based on how the participants approach it. Some crews might enjoy playing for 45-60 minutes and then save and come back later. 

That said, linear, railroaded missions will be possible with this framework. This just opens up possibilities for other mission writers with different goals.

> us commercial centers

I want to make it clear that commercial centers are totally welcome to use Thorium Nova, but our goal is primarily to create a great experience for friends and families playing in their living room. Quoting from [the vision document](https://github.com/Thorium-Sim/thorium-nova#project-vision):

> Thorium Nova's primary audience is casual gamers and hobbyists who want to run simulations for their friends and families. Thorium Nova might also be useful for brick-and-mortar space centers and simulators set up at conventions, but it must first be fun and usable for living room play.

> we also have normal operations that will require us to be able to limit where and how a crew can go.

Between writing the mission in a linear fashion and having a Flight Director who can redirect the crew, I think you'll have all the tools that you need to run missions in a similar fashion to how you're used to.
